### PR TITLE
SNOW-668660: Inefficient file lock for OCSP cache mechanism

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-668660

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [x] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

Reduce the number of lock and writing operations when parsing the `ocsp_response_cache.json`.
`ocsp_response_cache.json` contains nearly ~200 items, previously we perform at least 2 times of lock acquisition/release on each item for finding/deleting/updating cache.
This is unnecessary and could be optimized by batching the operations -- we acquire the lock at the beginning, then updating all the entries, then release the lock and write the cache back to file.

### CODE REVIEW

File: src/snowflake/connector/cache.py [hunk 1]
```diff
@@ -101,11 +100,7 @@ def _getitem(
             self._hit(k)
         return v
 
-    def _setitem(
-        self,
-        k: K,
-        v: V,
-    ) -> None:
+    def _setitem(self, k: K, v: V, **kwargs: Any) -> None:
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/cache.py:103]: Why did you add this?

```suggestion
    def _setitem(self, k: K, v: V) -> None:
```

File: src/snowflake/connector/cache.py [hunk 2]
```diff
@@ -86,7 +86,6 @@ def _getitem(
         should_record_hits: bool = True,
     ) -> V:
         """Non-locking version of __getitem__.
-
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/cache.py:89]: We usually leave an empty line between short and long descriptions.

File: src/snowflake/connector/cache.py [hunk 3]
```diff
@@ -407,6 +412,34 @@ def __init__(
         if os.path.exists(self.file_path):
             self._load()
 
+    def _getitem(
+        self,
+        k: K,
+        *,
+        should_record_hits: bool = True,
+    ) -> V:
+        """non-locking version of __getitem__
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/cache.py:421]:
```suggestion
        """Non-locking version of __getitem__.
        
```

File: src/snowflake/connector/cache.py [hunk 4]
```diff
@@ -187,17 +182,27 @@ def __delitem__(
         with self._lock:
             self._delitem(key)
 
+    def _contains(
+        self,
+        key: K,
+    ) -> bool:
+        """non-locking version of __contains__
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/cache.py:189]:
```suggestion
        """Non-locking version of __contains__.
        
```

File: src/snowflake/connector/cache.py [hunk 5]
```diff
@@ -407,6 +412,34 @@ def __init__(
...
+            if loaded:
+                t, v = self._cache[k]
+                expire_item = is_expired(t)
+            if expire_item:
+                # Raises KeyError
+                self._expire(k)
+        self._hit(k)
+        return v
+
     def __getitem__(self, k: K) -> V:
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/cache.py:443]: I think this function should reuse `self._getitem`, something like:
```python
 def __getitem__(self, k: K) -> V:
	with self._lock:
	    return self._getitem(k, v)
```

File: src/snowflake/connector/ocsp_snowflake.py [hunk 6]
```diff
@@ -695,14 +695,21 @@ def is_cache_fresh(current_time, ts):
         return current_time - OCSPCache.CACHE_EXPIRATION <= ts
 
     @staticmethod
-    def find_cache(ocsp, cert_id, subject, **kwargs):
+    def find_cache(
+        ocsp: SnowflakeOCSP, cert_id: CertId, subject: Certificate | None, **kwargs: Any
+    ):
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/ocsp_snowflake.py:700]: Return type-hint, please

File: src/snowflake/connector/ocsp_snowflake.py [hunk 7]
```diff
@@ -695,14 +695,21 @@ def is_cache_fresh(current_time, ts):
...
         subject_name = ocsp.subject_name(subject) if subject else None
         current_time = int(time.time())
         cache_key: tuple[bytes, bytes, bytes] = kwargs.get(
             "cache_key", ocsp.decode_cert_id_key(cert_id)
         )
-        if cache_key in OCSP_RESPONSE_VALIDATION_CACHE:
-            ocsp_response_validation_result = OCSP_RESPONSE_VALIDATION_CACHE[cache_key]
+        lock_cache = kwargs.get("lock_cache", True)
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/ocsp_snowflake.py:706]: If you want to use `kwargs` I have a hard requirement that you need to type-hint these lines explicitly
```suggestion
        lock_cache: bool = kwargs.get("lock_cache", True)
```

File: src/snowflake/connector/cache.py (hunk 1 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/cache.py]: it's no longer needed as I will use the `_update` method

File: src/snowflake/connector/cache.py (hunk 5 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/cache.py:435]: I'm afraid we can't, if you remember, the `__getitem__` implementation of `SFDictFileCache` has more acquire/release operations in the middle and I think that's for performance consideration.

File: src/snowflake/connector/ocsp_snowflake.py [hunk 8]
```diff
@@ -715,52 +722,28 @@ def find_cache(ocsp, cert_id, subject, **kwargs):
...
+                        ocsp, cert_id, cache_key=cache_key, lock_cache=lock_cache
+                    )
             except Exception as ex:
                 logger.debug(f"Could not validate cache entry {cert_id} {ex}")
             OCSPCache.CACHE_UPDATED = True
-        if subject_name:
-            logger.debug("not hit cache for subject: %s", subject_name)
+        except KeyError:
+            if subject_name:
+                logger.debug("not hit cache for subject: %s", subject_name)
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/ocsp_snowflake.py:733]:
```suggestion
                logger.debug(f"cache miss for subject: '{subject_name}'")
```

**Review by sfc-gh-mkeller — APPROVED** [review]
This is awesome, the code is super readable!

:ship: :ship:
